lib/prom/prom.go

```diff
@@ -0,0 +1,127 @@
+package prom
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+
+	vegeta "github.com/tsenart/vegeta/v12/lib"
+)
+
+// Metrics vegeta metrics observer with exposition as Prometheus metrics endpoint
+type Metrics struct {
+	requestSecondsHistogram *prometheus.HistogramVec
+	requestBytesInCounter   *prometheus.CounterVec
+	requestBytesOutCounter  *prometheus.CounterVec
+	requestFailCounter      *prometheus.CounterVec
+	srv                     http.Server
+	registry                *prometheus.Registry
+}
+
+// NewMetrics same as NewMetricsWithParams with default params:
+func NewMetrics() (*Metrics, error) {
+	return NewMetricsWithParams("http://0.0.0.0:8880")
+}
+
+// NewMetricsWithParams creates a new Prometheus Metrics to Observe attack results and expose metrics
+// For example, after using NewMetricsWithParams("http://0.0.0.0:8880"),
+// during an "attack" you can call "curl http://127.0.0.0:8880" to see current metrics.
+// This endpoint can be configured in scrapper section of your Prometheus server.
+func NewMetricsWithParams(bindURL string) (*Metrics, error) {
+
+	// parse bind url elements
+	p, err := url.Parse(bindURL)
+	if err != nil {
+		return nil, fmt.Errorf("Invalid bindURL %s. Must be in format 'http://0.0.0.0:8880'. err=%s", bindURL, err)
+	}
+	bindHost, bindPort, err := net.SplitHostPort(p.Host)
+	if err != nil {
+		return nil, fmt.Errorf("Invalid bindURL %s. Must be in format 'http://0.0.0.0:8880'. err=%s", bindURL, err)
+	}
+
+	pm := &Metrics{
+		registry: prometheus.NewRegistry(),
```

Comment:
I was trying to avoid mixing prometheus with attack.go at first, making it work only on cli so we don't change "attack" before we check the stability of prometheus metrics.

Reply:
Since `lib/prom` is a public package, we need to support library usage as well. We don't need to change `lib/attack.go`, only `attack.go`, where you'd instantiate the server with Prometheus handler. So this type `Metrics` would take in the `prometheus.Registry` and register it's metrics in it, but would not serve anything.

Reply:
I separated HTTP from metrics and started HTTP server from attack.go, but created an utility function in prom lib to help with this, so that we can use the same piece of code when creating the prom command in the future.

Reply:
Now it's more clear that someone can use the lib to prepare their own Prom Registry with Vegeta Metrics.